pynsot/commands/cmd_protocols.py

```diff
@@ -0,0 +1,472 @@
+
+
+# List
+@cli.group(invoke_without_command=True)
+@click.option(
+    '-a',
+    '--attributes',
+    metavar='ATTRS',
+    help='A key/value pair attached to this Protocol (format: key=value).',
+    multiple=True,
+)
+@click.option(
+    '-u',
+    '--auth_string',
+    metavar='AUTH_STRING',
+    default='',
+    help='The authentication string (such as MD5 sum).',
+)
+@click.option(
+    '-c',
+    '--circuit',
+    metavar='CIRCUIT',
+    help='The circuit that this protocol is running over.',
+)
+@click.option(
+    '-d',
+    '--delimited',
+    is_flag=True,
+    help='Display set query results separated by commas vs. newlines.',
+    default=False,
+    show_default=True,
+)
+@click.option(
+    '-e',
+    '--description',
+    metavar='DESCRIPTION',
+    type=str,
+    help='Filter by Protocols matching this description.',
+)
+@click.option(
+    '-D',
+    '--device',
+    metavar='DEVICE',
+    type=types.NATURAL_KEY,
+    help=(
+        'Unique ID of the Device to which this Protocol is '
+        'running on.'
+    ),
+)
+@click.option(
+    '-g',
+    '--grep',
+    is_flag=True,
+    help='Display list results in a grep-friendly format.',
+    default=False,
+    show_default=True,
+)
+@click.option(
+    '-I',
+    '--id',
+    metavar='ID',
+    help='Unique ID of the Protocol being retrieved.',
+)
+@click.option(
+    '-i',
+    '--interface',
+    metavar='INTERFACE',
+    type=types.NATURAL_KEY,
+    help=(
+        'The Interface this Protocol is running on. Either interface'
+        'or circuit must be populated.'
+    ),
+)
+@click.option(
+    '-q',
+    '--query',
+    metavar='QUERY',
+    help='Perform a set query using Attributes and output matching Protocols.'
+)
+@click.option(
+    '-s',
+    '--site-id',
+    metavar='SITE_ID',
+    help='Unique ID of the Site this Protocol is under.',
+    callback=callbacks.process_site_id,
+)
+@click.option(
+    '-t',
+    '--type',
+    metavar='TYPE',
+    type=types.NATURAL_KEY,
+    help='The protocol type',
+    required=True,
+)
+@click.pass_context
+def list(ctx, attributes, auth_string, circuit, delimited, description, device, grep, id,
+ 		 interface, query, site_id, type):
+    """
+    List existing Protocols for a Site.
+
+    You must provide a Site ID using the -s/--site-id option.
+
+    You must provide the protocol type.
+
+    When listing Protocols, all objects are displayed by default. You
+    optionally may lookup a single Protocols by ID using the -i/--id option.
+
+    You may look up the protocols for a single Device using -D/--device which
+    can be either a hostname or ID for a Device.
+    """
+    data = ctx.params
+    data.pop('delimited')  # We don't want this going to the server.
+
+    # If we provide ID, show more fields
+    if id is not None:
+        display_fields = VERBOSE_FIELDS
+    else:
+        display_fields = DISPLAY_FIELDS
+
+    # FIXME(jathan): If it's not a digit, it's a hostname? This is a hack for
```

Comment:
It appears you duplicated this from the `interfaces` command, but this is not relevant for `Protocol` so it's not accurate to include this here.  `Protocol` doesn't have a `device_hostname` field. This was a hack that was in place for filtering Interfaces by their device before we implemented native natural key filtering on the server side. It's actually something we could probably remove in a future patch.